### PR TITLE
Graduate the JobSuccessPolicy to beta

### PR DESCRIPTION
**Pull request by tenzen-y**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

Beta Graduation PR for the `JobSuccessPolicy`

#### REVIEW STEPS

This PR is part of the JobSuccessPolicy Beta Graduation PR.
So, please take a look in the following steps:

1. https://github.com/kubernetes/kubernetes/pull/126074
2. https://github.com/kubernetes/kubernetes/pull/126075
3. This PR (https://github.com/kubernetes/kubernetes/pull/126064/commits/c814701c9977bee37bd3e3880be1ffc9313bbab0).
4. https://github.com/kubernetes/kubernetes/pull/126067

#### Which issue(s) this PR fixes:

Part-of: https://github.com/kubernetes/enhancements/issues/3998

#### Special notes for your reviewer:

Beta Graduation Criteria from [KEP-3998](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3998-job-success-completion-policy#beta):

- [x] E2E tests passed as designed in [TestPlan](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3998-job-success-completion-policy#test-plan): https://github.com/kubernetes/kubernetes/pull/126067
- [x] Introduced a new `job_succeeded_total` metric in [Monitoring Requirements](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3998-job-success-completion-policy#monitoring-requirements): https://github.com/kubernetes/kubernetes/pull/126075
- [x] Feature is enabled by default: This PR.
- [x] Address all issues reported by users:
  - [x] https://github.com/kubernetes/kubernetes/pull/125429 

Additionally, the PR introducing a new metric (https://github.com/kubernetes/kubernetes/pull/126075) depends on a new `CompletionsReached` reason (https://github.com/kubernetes/kubernetes/pull/126074) for the `SuccessCriteriaMet` condition, which means this PR depends on https://github.com/kubernetes/kubernetes/pull/126074 as well.

#### Does this PR introduce a user-facing change?

```release-note
Graduate Job SuccessPolicy to Beta.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3998-job-success-completion-policy
```

### CODE REVIEW

**Comment by tenzen-y** [comment]
Please take a look.
/assign @mimowo @soltysh @alculquicondor @atiratree

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -985,10 +985,16 @@ func (jm *Controller) newFailureCondition(reason, message string) *batch.JobCond
 
 func (jm *Controller) newSuccessCondition() *batch.JobCondition {
 	cType := batch.JobComplete
+	var reason, message string
 	if delayTerminalCondition() {
 		cType = batch.JobSuccessCriteriaMet
+
+		// To avoid the breaking the existing Completion condition, the CompletionsReached reason and message are set
+		// only when the condition type is SuccessCriteriaMet.
+		reason = batch.JobReasonCompletionsReached
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:994]: Let's move this discussion here: https://github.com/kubernetes/kubernetes/pull/126074#discussion_r1677391721

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -1198,19 +1204,28 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
...
 		if hasSuccessCriteriaMetCondition(jobCtx.job) == nil {
 			jobCtx.job.Status.Conditions = append(jobCtx.job.Status.Conditions, *jobCtx.finishedCondition)
-			needsFlush = true
+			needsIncrementSucceededMetric = true
 		}
 
+		// To avoid the breaking the existing Completion condition, the SuccessCriteriaMet reason and message are propagated
+		// only when the SuccessCriteriaMet condition reason is not CompletionsReached.
+		var reason, message string
+		if jobCtx.finishedCondition.Reason != batch.JobReasonCompletionsReached {
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:1218]: Let's move this discussion here: https://github.com/kubernetes/kubernetes/pull/126074#discussion_r1677392103

File: pkg/controller/job/job_controller.go [hunk 3]
```diff
@@ -1429,6 +1455,22 @@ func (jm *Controller) recordJobFinished(job *batch.Job, finishedCond *batch.JobC
...
+		return
+	}
+	var metricReason string
+	switch successCriteriaMet.Reason {
+	case batch.JobReasonSuccessPolicy:
+		metricReason = "JobSuccessPolicy"
+	case batch.JobReasonCompletionsReached:
+		metricReason = "Completions"
+	}
+	jm.recorder.Event(job, v1.EventTypeNormal, "SuccessCriteriaMet", "Job met success criteria")
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:1470]: Let's move this discussion here: https://github.com/kubernetes/kubernetes/pull/126075#discussion_r1677410033

File: pkg/controller/job/job_controller.go [hunk 4]
```diff
@@ -1429,6 +1455,22 @@ func (jm *Controller) recordJobFinished(job *batch.Job, finishedCond *batch.JobC
...
+	metrics.JobSucceededTotal.WithLabelValues(metricReason).Inc()
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:1471]: Let's move this discussion here: https://github.com/kubernetes/kubernetes/pull/126075#discussion_r1677411216

File: pkg/controller/job/job_controller.go [hunk 5]
```diff
@@ -1429,6 +1455,22 @@ func (jm *Controller) recordJobFinished(job *batch.Job, finishedCond *batch.JobC
 	return true
 }
 
+func (jm *Controller) recordJobSucceeded(job *batch.Job) {
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:1458]: Let's move this discussion here: https://github.com/kubernetes/kubernetes/pull/126075#discussion_r1677412171

**Comment by tenzen-y** [comment]
@mimowo Thank you for the review. I moved your comments to dedicated PR.

**Comment by mimowo** [comment]
> Additionally, the PR introducing a new metric (#126075) depends on a new `CompletionsReached` reason (#126074) for the `SuccessCriteriaMet` condition, which means this PR depends on #126074 as well.

There are 4 small PRs involved:
- https://github.com/kubernetes/kubernetes/pull/126075
- https://github.com/kubernetes/kubernetes/pull/126074
- https://github.com/kubernetes/kubernetes/pull/126064
- https://github.com/kubernetes/kubernetes/pull/126067

I would consider combining the PRs into one, because:
- there are dependencies, conceptual (e.g. decision about extending the reasons holds the need for new metric) and technical (e.g. graduation to beta holds e2e tests) between them. The conceptual dependencies make the discussion scattered
- the combined PR would still be relatively small (imo)
- we only have 7 days remaining, and there is non-negligible overhead on merging one-by-one (merge itself and rebase)

WDYT @soltysh @alculquicondor @tenzen-y ?

**Comment by soltysh** [comment]
> WDYT @soltysh @alculquicondor @tenzen-y ?

I wouldn't object into having just a single PR, iirc https://github.com/kubernetes/kubernetes/pull/126067 has all the bits in place, so maybe just focusing on landing that as is?

**Comment by tenzen-y** [comment]
> > Additionally, the PR introducing a new metric (#126075) depends on a new `CompletionsReached` reason (#126074) for the `SuccessCriteriaMet` condition, which means this PR depends on #126074 as well.
> 
> There are 4 small PRs involved:
> 
> * [Job: Introduce a new job_suceeded_total metric #126075](https://github.com/kubernetes/kubernetes/pull/126075)
> * [Job: Add the CompletionsReached reason to the SuccessCriteriaMet condition #126074](https://github.com/kubernetes/kubernetes/pull/126074)
> * [Graduate the JobSuccessPolicy to beta #126064](https://github.com/kubernetes/kubernetes/pull/126064)
> * [Implement E2E tests for the JobSuccessPolicy #126067](https://github.com/kubernetes/kubernetes/pull/126067)
> 
> I would consider combining the PRs into one, because:
> 
> * there are dependencies, conceptual (e.g. decision about extending the reasons holds the need for new metric) and technical (e.g. graduation to beta holds e2e tests) between them. The conceptual dependencies make the discussion scattered
> * the combined PR would still be relatively small (imo)
> * we only have 7 days remaining, and there is non-negligible overhead on merging one-by-one (merge itself and rebase)
> 
> WDYT @soltysh @alculquicondor @tenzen-y ?

Thank you for raising the suggestion! I opened several PRs so that we can revert specific commits easily when we face the issues something. (I would hope that there are not bugs and regressions in this feature, although).

However, I don't have any strong objections to focusing on a single PR. Let me try organizing all PRs.

**Comment by alculquicondor** [comment]
Please let me know which PR I should be reviewing.

**Comment by tenzen-y** [comment]
We decided to focus only on https://github.com/kubernetes/kubernetes/pull/126067.
/close.

**Pull request closed without merging**